File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
//  set delay     
    let delayCounter = 0, fast = head, slow = head, prev = null;
    
    if(!head.next) {
        return head.next;
    } 
    
//  fast starts and slow starts after delay counter equals the delay
    while(fast) {
        if(delayCounter >= n) {
            prev = slow;
            slow = slow.next;
        }
        fast = fast.next;
        delayCounter++;
    }
    
    if(prev === null) {
        return head = head.next;
    }
    prev.next = slow.next;
    return head;
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
//         define slow and fast pointers
        ListNode slow = head, fast = head;
        
        for(int i = 0; i < n; i++) {
            fast = fast.next;
        }
        
//         edge case
        if(fast == null) return head.next;
        
//         iterate
        while(fast.next != null) {
            fast = fast.next;
            slow = slow.next;
        }
        
//         remove the slow pointer's next pointer
        slow.next = slow.next.next;
        
        return head;
    }
}